### PR TITLE
KOGITO-8391 Kogito example of Listener for DT events

### PR DESCRIPTION
**Pull request by tarilabs**
ref https://issues.redhat.com/browse/KOGITO-8391

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [kogito-examples] tests</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [kogito-examples] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [kogito-examples] quarkus-main</b>

- for <b>quarkus lts checks</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins run quarkus-lts</b>

- for a <b>specific quarkus lts check</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins (re)run [kogito-examples] quarkus-lts</b>
 
- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [kogito-examples] native</b>

- for <b>mandrel checks</b>  
  Run native checks against Mandrel image
  Please add comment: <b>Jenkins run mandrel</b>

- for a <b>specific mandrel check</b>  
  Run native checks against Mandrel image  
  Please add comment: <b>Jenkins (re)run [kogito-examples] mandrel</b>

- for <b>mandrel lts checks</b>  
  Run native checks against Mandrel image and quarkus lts branch
  Please add comment: <b>Jenkins run mandrel-lts</b>

- for a <b>specific mandrel lts check</b>  
  Run native checks against Mandrel image and quarkus lts branch 
  Please add comment: <b>Jenkins (re)run [kogito-examples] mandrel-lts</b>
 
</details>

### CODE REVIEW

File: kogito-quarkus-examples/dmn-listener-dtable/src/main/java/org/kie/kogito/dmn/quarkus/example/dtlistener/ExampleDMNRuntimeEventListener.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ */
+@ApplicationScoped
+public class ExampleDMNRuntimeEventListener extends DefaultDMNRuntimeEventListener {
+    // using a ConcurrentLinkedQueue to avoid ConcurrentModificationException
+    private Queue<AfterEvaluateDecisionTableEvent> events = new ConcurrentLinkedQueue<>();
+
+    @Override
+    public void afterEvaluateDecisionTable(AfterEvaluateDecisionTableEvent event) {
+        // Maintaining listener callback code as small as possible.
+        events.add(event);
```
Comment (tarilabs) [inline kogito-quarkus-examples/dmn-listener-dtable/src/main/java/org/kie/kogito/dmn/quarkus/example/dtlistener/ExampleDMNRuntimeEventListener.java:37]: make the listener callback code as small as possible

File: kogito-quarkus-examples/dmn-listener-dtable/src/main/java/org/kie/kogito/dmn/quarkus/example/dtlistener/PeriodicJobBean.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    private static final Logger LOG = LoggerFactory.getLogger(PeriodicJobBean.class);
+
+    @Inject
+    ExampleDMNRuntimeEventListener listener;
+
+    @Inject
+    DecisionModels models; // can't use yet Incubation API here as I need access to DMNModel
+
+    @Scheduled(every = "5s")
+    public void logEvents() {
```
Comment (tarilabs) [inline kogito-quarkus-examples/dmn-listener-dtable/src/main/java/org/kie/kogito/dmn/quarkus/example/dtlistener/PeriodicJobBean.java:49]: process the events asynchronously

File: kogito-quarkus-examples/dmn-listener-dtable/src/main/java/org/kie/kogito/dmn/quarkus/example/dtlistener/PeriodicJobBean.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        for (AfterEvaluateDecisionTableEvent event : listener.getEvents()) {
+            DecisionTable dt = models.getDecisionModel("myNS", "dtevent").getDMNModel().getDefinitions()
+                    .findAllChildren(DecisionTable.class).stream()
+                    .filter(t -> t.getId().equals(event.getDecisionTableId()))
+                    .findFirst().orElseThrow(IllegalStateException::new); // only one having that ID, checked during build time by kie-dmn-validation
+            LOG.info("Decision Table ({}) having a total of {} rows, during evaluation matched rows: {}, and selected rows: {}.",
+                    event.getDecisionTableId(),
+                    dt.getRule().size(), // from the original DT definition
+                    event.getMatches(), // which rules have matched during evaluation
+                    event.getSelected()); // which rules have been selected from matching set, by hit policy
```
Comment (tarilabs) [inline kogito-quarkus-examples/dmn-listener-dtable/src/main/java/org/kie/kogito/dmn/quarkus/example/dtlistener/PeriodicJobBean.java:59]: > Do you have any examples?

providing example.

File: kogito-quarkus-examples/dmn-listener-dtable/src/test/java/org/kie/kogito/dmn/quarkus/example/dtlistener/DecisionTableTest.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import static org.awaitility.Awaitility.await;
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.hasKey;
+
+/**
+ * This test will remain just-and-only a `@QuarkusTest` given the need of CDI injection to verify assertions.
+ * This test does not need an IT equivalent, simply because the logic provided by the listener is not exposed as REST, for instance.
+ */
+@QuarkusTest
+public class DecisionTableTest {
```
Comment (tarilabs) [inline kogito-quarkus-examples/dmn-listener-dtable/src/test/java/org/kie/kogito/dmn/quarkus/example/dtlistener/DecisionTableTest.java:39]: as the comments says, per the no-need for IT variant.

**Review by tarilabs — COMMENTED** [review]
I am planning to use this example to reply here: https://groups.google.com/g/kogito-development/c/oFj4zV-Oy_s/m/S5jQaPgmAQAJ

**Comment by tarilabs** [comment]
I forgot to wire the module in the parent pom 🤦‍♂️ I will amend it!

**Review by mariofusco — APPROVED** [review]

**Review by danielezonca — APPROVED** [review]

**Pull request merged**